### Подключение модулей

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np

### Загрузка данных из облака в датафреймы

In [2]:
assessments = pd.read_csv('assessments.csv')
courses = pd.read_csv('courses.csv')
studentAssessment = pd.read_csv('studentAssessment.csv')
studentRegistration = pd.read_csv('studentRegistration.csv')

### Изучение данных
#### assessments - содержит информацию об оценках в тесте. Обычно каждый предмет в семестре включает ряд тестов с оценками, за которыми следует заключительный экзаменационный тест (экзамен).
code_module — идентификационный код предмета.

code_presentation — семестр (Идентификационный код).

id_assessment — тест (Идентификационный номер ассессмента).

assessment_type — тип теста. Существуют три типа оценивания: оценка преподавателя (TMA), компьютерная оценка (СМА), экзамен по курсу (Exam).

date — информация об окончательной дате сдачи теста. Рассчитывается как количество дней с момента начала семестра. Дата начала семестра имеет номер 0 (ноль).

weight — вес теста в % в оценке за курс. Обычно экзамены рассматриваются отдельно и имеют вес 100%; сумма всех остальных оценок составляет 100%.

In [3]:
assessments.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

In [4]:
assessments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    object 
 1   code_presentation  206 non-null    object 
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    object 
 4   date               195 non-null    float64
 5   weight             206 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 9.8+ KB


In [5]:
assessments.nunique()

code_module            7
code_presentation      4
id_assessment        206
assessment_type        3
date                  74
weight                24
dtype: int64

In [6]:
assessments.describe()

id_assessment        date      weight
count     206.000000  195.000000  206.000000
mean    26473.975728  145.005128   20.873786
std     10098.625521   76.001119   30.384224
min      1752.000000   12.000000    0.000000
25%     15023.250000   71.000000    0.000000
50%     25364.500000  152.000000   12.500000
75%     34891.750000  222.000000   24.250000
max     40088.000000  261.000000  100.000000

In [7]:
assessments.duplicated().sum()

0

In [8]:
assessments.isnull().sum()

code_module           0
code_presentation     0
id_assessment         0
assessment_type       0
date                 11
weight                0
dtype: int64

In [9]:
assessments.loc[assessments['date'].isna()]

code_module code_presentation  id_assessment assessment_type  date  weight
5           AAA             2013J           1757            Exam   NaN   100.0
11          AAA             2014J           1763            Exam   NaN   100.0
23          BBB             2013B          14990            Exam   NaN   100.0
35          BBB             2013J          15002            Exam   NaN   100.0
47          BBB             2014B          15014            Exam   NaN   100.0
53          BBB             2014J          15025            Exam   NaN   100.0
62          CCC             2014B          24290            Exam   NaN   100.0
63          CCC             2014B          40087            Exam   NaN   100.0
72          CCC             2014J          24299            Exam   NaN   100.0
73          CCC             2014J          40088            Exam   NaN   100.0
108         DDD             2014J          25368            Exam   NaN   100.0

In [10]:
assessments[assessments.assessment_type == 'Exam']

code_module code_presentation  id_assessment assessment_type   date  \
5           AAA             2013J           1757            Exam    NaN   
11          AAA             2014J           1763            Exam    NaN   
23          BBB             2013B          14990            Exam    NaN   
35          BBB             2013J          15002            Exam    NaN   
47          BBB             2014B          15014            Exam    NaN   
53          BBB             2014J          15025            Exam    NaN   
62          CCC             2014B          24290            Exam    NaN   
63          CCC             2014B          40087            Exam    NaN   
72          CCC             2014J          24299            Exam    NaN   
73          CCC             2014J          40088            Exam    NaN   
87          DDD             2013B          25340            Exam  240.0   
94          DDD             2013J          25354            Exam  261.0   
101         DDD             2014B          25361            Exam  241.0   
108         DDD             2014J          25368            Exam    NaN   
113         EEE             2013J          30713            Exam  235.0   
118         EEE             2014B          30718            Exam  228.0   
123         EEE             2014J          30723            Exam  235.0   
136         FFF             2013B          34872            Exam  222.0   
149         FFF             2013J          34885            Exam  236.0   
162         FFF             2014B          34898            Exam  227.0   
175         FFF             2014J          34911            Exam  241.0   
185         GGG             2013J          37424            Exam  229.0   
195         GGG             2014B          37434            Exam  222.0   
205         GGG             2014J          37444            Exam  229.0   

     weight  
5     100.0  
11    100.0  
23    100.0  
35    100.0  
47    100.0  
53    100.0  
62    100.0  
63    100.0  
72    100.0  
73    100.0  
87    100.0  
94    100.0  
101   100.0  
108   100.0  
113   100.0  
118   100.0  
123   100.0  
136   100.0  
149   100.0  
162   100.0  
175   100.0  
185   100.0  
195   100.0  
205   100.0

#### Зпишем в date -1 вместо NaN. Это будет признаком того, что дата экзамена не определена.

In [11]:
assessments.date = assessments.date.fillna(-1)

assessments.isnull().sum()

code_module          0
code_presentation    0
id_assessment        0
assessment_type      0
date                 0
weight               0
dtype: int64

#### courses - содержит список предметов по семестрам.
code_module — предмет (идентификационный код).

code_presentation — семестр (идентификационный код).

module_presentation_length — продолжительность семестра в днях.

In [12]:
courses.head()

code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [13]:
courses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   code_module                 22 non-null     object
 1   code_presentation           22 non-null     object
 2   module_presentation_length  22 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 656.0+ bytes


In [14]:
courses.nunique()

code_module                   7
code_presentation             4
module_presentation_length    7
dtype: int64

In [15]:
courses.describe()

module_presentation_length
count                   22.000000
mean                   255.545455
std                     13.654677
min                    234.000000
25%                    241.000000
50%                    261.500000
75%                    268.000000
max                    269.000000

In [16]:
courses.duplicated().sum()

0

#### studentAssessment - содержит результаты тестов студентов. Если учащийся не отправляет работу на оценку, результат не записывается в таблицу.
id_assessment — тест (идентификационный номер).

id_student — идентификационный номер студента.

date_submitted — дата сдачи теста студентом, измеряемая как количество дней с начала семестра.

is_banked — факт перезачета теста с прошлого семестра (иногда курсы перезачитывают студентам, вернувшимся из академического отпуска).

score — оценка учащегося в этом тесте. Диапазон составляет от 0 до 100. Оценка ниже 40 неудачная/неуспешная сдача теста.

In [17]:
studentAssessment.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [18]:
studentAssessment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id_assessment   173912 non-null  int64  
 1   id_student      173912 non-null  int64  
 2   date_submitted  173912 non-null  int64  
 3   is_banked       173912 non-null  int64  
 4   score           173739 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 6.6 MB


In [19]:
studentAssessment.nunique()

id_assessment       188
id_student        23369
date_submitted      312
is_banked             2
score               101
dtype: int64

In [20]:
studentAssessment.describe()

id_assessment    id_student  date_submitted      is_banked  \
count  173912.000000  1.739120e+05   173912.000000  173912.000000   
mean    26553.803556  7.051507e+05      116.032942       0.010977   
std      8829.784254  5.523952e+05       71.484148       0.104194   
min      1752.000000  6.516000e+03      -11.000000       0.000000   
25%     15022.000000  5.044290e+05       51.000000       0.000000   
50%     25359.000000  5.852080e+05      116.000000       0.000000   
75%     34883.000000  6.344980e+05      173.000000       0.000000   
max     37443.000000  2.698588e+06      608.000000       1.000000   

               score  
count  173739.000000  
mean       75.799573  
std        18.798107  
min         0.000000  
25%        65.000000  
50%        80.000000  
75%        90.000000  
max       100.000000

In [21]:
studentAssessment.duplicated().sum()

0

In [22]:
studentAssessment.isnull().sum()

id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64

In [23]:
studentAssessment.loc[studentAssessment['score'].isna()]

id_assessment  id_student  date_submitted  is_banked  score
215              1752      721259              22          0    NaN
937              1754      260355             127          0    NaN
2364             1760     2606802             180          0    NaN
3358            14984      186780              77          0    NaN
3914            14984      531205              26          0    NaN
...               ...         ...             ...        ...    ...
148929          34903      582670             241          0    NaN
159251          37415      610738              87          0    NaN
166390          37427      631786             221          0    NaN
169725          37435      648110              62          0    NaN
170103          37435      480914              49          0    NaN

[173 rows x 5 columns]

#### Запишем в score 0 вместо NaN. Это будет признаком того, что студент не сдал тест.

In [24]:
studentAssessment.score = studentAssessment.score.fillna(0)

studentAssessment.isnull().sum()

id_assessment     0
id_student        0
date_submitted    0
is_banked         0
score             0
dtype: int64

#### studentRegistration - содержит информацию о времени, когда студент зарегистрировался для прохождения курса в семестре.
code_module — предмет (идентификационный код).

code_presentation — семестр (идентификационный код)

id_student — идентификационный номер студента.

date_registration — дата регистрации студента. Это количество дней, измеренное от начала семестра (например, отрицательное значение -30 означает, что студент зарегистрировался на прохождение курса за 30 дней до его начала).

date_unregistration — дата отмены регистрации студента с предмета. У студентов, окончивших курс, это поле остается пустым.

In [25]:
studentRegistration.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

In [26]:
studentRegistration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   code_module          32593 non-null  object 
 1   code_presentation    32593 non-null  object 
 2   id_student           32593 non-null  int64  
 3   date_registration    32548 non-null  float64
 4   date_unregistration  10072 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 1.2+ MB


In [27]:
studentRegistration.nunique()

code_module                7
code_presentation          4
id_student             28785
date_registration        332
date_unregistration      416
dtype: int64

In [28]:
studentRegistration.describe()

id_student  date_registration  date_unregistration
count  3.259300e+04       32548.000000         10072.000000
mean   7.066877e+05         -69.411300            49.757645
std    5.491673e+05          49.260522            82.460890
min    3.733000e+03        -322.000000          -365.000000
25%    5.085730e+05        -100.000000            -2.000000
50%    5.903100e+05         -57.000000            27.000000
75%    6.444530e+05         -29.000000           109.000000
max    2.716795e+06         167.000000           444.000000

In [29]:
studentRegistration.duplicated().sum()

0

In [30]:
studentRegistration.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64

In [31]:
studentRegistration.loc[studentRegistration['date_registration'].isna()]

code_module code_presentation  id_student  date_registration  \
2344          BBB             2013B      630346                NaN   
2538          BBB             2013J       57369                NaN   
2759          BBB             2013J      342678                NaN   
5356          BBB             2014B      582496                NaN   
5490          BBB             2014B      607646                NaN   
5573          BBB             2014B      614270                NaN   
6295          BBB             2014B     2409808                NaN   
6305          BBB             2014B     2439442                NaN   
8307          BBB             2014J      694001                NaN   
8975          CCC             2014B      394791                NaN   
9523          CCC             2014B      575419                NaN   
9811          CCC             2014B      604480                NaN   
9853          CCC             2014B      607703                NaN   
10388         CCC             2014B     1600062                NaN   
11249         CCC             2014J      544271                NaN   
12545         CCC             2014J      680333                NaN   
12893         CCC             2014J     1777834                NaN   
13135         DDD             2013B      128965                NaN   
13503         DDD             2013B      464615                NaN   
14392         DDD             2013B     2707979                NaN   
14393         DDD             2013B     2710343                NaN   
14556         DDD             2013J      250456                NaN   
14876         DDD             2013J      470119                NaN   
15214         DDD             2013J      550142                NaN   
15968         DDD             2013J      604398                NaN   
16092         DDD             2013J      835603                NaN   
16862         DDD             2014B      591648                NaN   
16911         DDD             2014B      604480                NaN   
17559         DDD             2014B     2710343                NaN   
18249         DDD             2014J      604632                NaN   
18433         DDD             2014J      636590                NaN   
19362         DDD             2014J     2716795                NaN   
19773         EEE             2013J      568751                NaN   
20413         EEE             2013J     2681277                NaN   
22335         FFF             2013B      107955                NaN   
23796         FFF             2013B     2102658                NaN   
23818         FFF             2013B     2288493                NaN   
25790         FFF             2013J      605108                NaN   
26333         FFF             2014B      323688                NaN   
26576         FFF             2014B      535460                NaN   
26803         FFF             2014B      598362                NaN   
26904         FFF             2014B      612254                NaN   
27550         FFF             2014B     1290731                NaN   
29974         FFF             2014J     2371481                NaN   
30058         FFF             2014J     2702660                NaN   

       date_unregistration  
2344                   NaN  
2538                  -1.0  
2759                 -33.0  
5356                -126.0  
5490                 -38.0  
5573                -142.0  
6295                -109.0  
6305                -149.0  
8307                 -36.0  
8975                 -61.0  
9523                 -24.0  
9811                -168.0  
9853                -156.0  
10388               -152.0  
11249                -29.0  
12545                -65.0  
12893                  NaN  
13135                -24.0  
13503               -116.0  
14392                  NaN  
14393                  NaN  
14556                -40.0  
14876                -75.0  
15214               -196.0  
15968                 -1.0  
16092                -81.0  
16

In [32]:
studentRegistration.loc[studentRegistration['date_registration'].isna() & \
                        ~studentRegistration['date_unregistration'].isna()]

code_module code_presentation  id_student  date_registration  \
2538          BBB             2013J       57369                NaN   
2759          BBB             2013J      342678                NaN   
5356          BBB             2014B      582496                NaN   
5490          BBB             2014B      607646                NaN   
5573          BBB             2014B      614270                NaN   
6295          BBB             2014B     2409808                NaN   
6305          BBB             2014B     2439442                NaN   
8307          BBB             2014J      694001                NaN   
8975          CCC             2014B      394791                NaN   
9523          CCC             2014B      575419                NaN   
9811          CCC             2014B      604480                NaN   
9853          CCC             2014B      607703                NaN   
10388         CCC             2014B     1600062                NaN   
11249         CCC             2014J      544271                NaN   
12545         CCC             2014J      680333                NaN   
13135         DDD             2013B      128965                NaN   
13503         DDD             2013B      464615                NaN   
14556         DDD             2013J      250456                NaN   
14876         DDD             2013J      470119                NaN   
15214         DDD             2013J      550142                NaN   
15968         DDD             2013J      604398                NaN   
16092         DDD             2013J      835603                NaN   
16862         DDD             2014B      591648                NaN   
16911         DDD             2014B      604480                NaN   
18249         DDD             2014J      604632                NaN   
18433         DDD             2014J      636590                NaN   
19362         DDD             2014J     2716795                NaN   
19773         EEE             2013J      568751                NaN   
20413         EEE             2013J     2681277                NaN   
22335         FFF             2013B      107955                NaN   
23818         FFF             2013B     2288493                NaN   
25790         FFF             2013J      605108                NaN   
26333         FFF             2014B      323688                NaN   
26576         FFF             2014B      535460                NaN   
26803         FFF             2014B      598362                NaN   
26904         FFF             2014B      612254                NaN   
27550         FFF             2014B     1290731                NaN   
29974         FFF             2014J     2371481                NaN   
30058         FFF             2014J     2702660                NaN   

       date_unregistration  
2538                  -1.0  
2759                 -33.0  
5356                -126.0  
5490                 -38.0  
5573                -142.0  
6295                -109.0  
6305                -149.0  
8307                 -36.0  
8975                 -61.0  
9523                 -24.0  
9811                -168.0  
9853                -156.0  
10388               -152.0  
11249                -29.0  
12545                -65.0  
13135                -24.0  
13503               -116.0  
14556                -40.0  
14876                -75.0  
15214               -196.0  
15968                 -1.0  
16092                -81.0  
16862               -213.0  
16911               -168.0  
18249                 12.0  
18433               -163.0  
19362                  5.0  
19773                146.0  
20413               -166.0  
22335                -40.0  
23818                 -2.0  
25790                -46.0  
26333               -107.0  
26576               -365.0  
26803               -119.0  
26904                -88.0  
27550                -82.0  
29974                -53.0  
30058                  9.0

#### Уберём те строки, где дата регистрации не указана и курс отменён (дата отмены регистрации не NaN). Такие наблюдения нам точно не нужны.

In [33]:
studentRegistration = studentRegistration.drop(studentRegistration[studentRegistration['date_registration'].isna() & \
                        ~studentRegistration['date_unregistration'].isna()].index)

studentRegistration.loc[studentRegistration['date_registration'].isna() & \
                        ~studentRegistration['date_unregistration'].isna()]

Empty DataFrame
Columns: [code_module, code_presentation, id_student, date_registration, date_unregistration]
Index: []

#### Запишем в date_registration вместо оставшихся NaN медианное значение. Затем запишем в date_unregistration 500 вместо NaN (500 это больше максимума, взятого из describe). Это будет признаком того, что студент окончил курс.

In [34]:
studentRegistration.date_registration = studentRegistration.date_registration \
                                    .fillna(studentRegistration.date_registration.median())

studentRegistration.date_unregistration = studentRegistration.date_unregistration.fillna(500)

studentRegistration.isnull().sum()

code_module            0
code_presentation      0
id_student             0
date_registration      0
date_unregistration    0
dtype: int64

In [35]:
studentRegistration.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                500.0  
1                500.0  
2                 12.0  
3                500.0  
4                500.0

### 0. Оценив задания, я решил добавить новое поле и объединить датафреймы:

#### Введём понятие "курс". Курс - это сущность, объединяющая предмет и семестр.
#### Добавим поле course_id в courses.

In [36]:

courses['course_id'] = courses['code_module'] + courses['code_presentation']

#### Присоединяю courses к assessments, получаю assessments_ext:

In [37]:
assessments_ext = assessments.merge(courses, how = 'inner', on = ['code_module', 'code_presentation'])

#### Затем присоединяю assessments_ext к studentAssessment, получаю studentAssessment_ext

In [38]:
studentAssessment_ext = studentAssessment.merge(assessments_ext, how = 'inner', on = 'id_assessment')

### 1. Сколько студентов успешно сдали только один курс? (Успешная сдача — это зачёт по курсу на экзамене).

#### В данном случае работаем с датафреймом studentAssessment_ext

In [39]:
studentAssessment_ext.head()

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0           1752       11391              18          0   78.0         AAA   
1           1752       28400              22          0   70.0         AAA   
2           1752       31604              17          0   72.0         AAA   
3           1752       32885              26          0   69.0         AAA   
4           1752       38053              19          0   79.0         AAA   

  code_presentation assessment_type  date  weight  module_presentation_length  \
0             2013J             TMA  19.0    10.0                         268   
1             2013J             TMA  19.0    10.0                         268   
2             2013J             TMA  19.0    10.0                         268   
3             2013J             TMA  19.0    10.0                         268   
4             2013J             TMA  19.0    10.0                         268   

  course_id  
0  AAA2013J  
1  AAA2013J  
2  AAA2013J  
3  AAA2013J  
4  AAA2013J

#### В условии успешная сдача - это зачёт на экзамне. Зачёт получают при score не меньше 40. Группируем по студенту, у каждого студента считаем количество уникальных курсов, берём из полученных значений только единицы (условие - сдал только один курс), считаем размер датафрейма, выводим количество строк.

In [40]:
studentAssessment_ext.query("assessment_type == 'Exam' and score >= 40") \
    .groupby('id_student', as_index = False)['course_id'].nunique() \
    .query('course_id == 1').shape[0]

3802

#### Ответ: 3802 студента успешно сдали только один курс.

### 2. Выяви самый сложный и самый простой экзамен: найди курсы и экзамены в рамках курса, которые обладают самой низкой и самой высокой завершаемостью*.
*завершаемость = кол-во успешных экзаменов / кол-во всех попыток сдать экзамен

#### Работаем с studentAssessment_ext

In [41]:
studentAssessment_ext.head()

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0           1752       11391              18          0   78.0         AAA   
1           1752       28400              22          0   70.0         AAA   
2           1752       31604              17          0   72.0         AAA   
3           1752       32885              26          0   69.0         AAA   
4           1752       38053              19          0   79.0         AAA   

  code_presentation assessment_type  date  weight  module_presentation_length  \
0             2013J             TMA  19.0    10.0                         268   
1             2013J             TMA  19.0    10.0                         268   
2             2013J             TMA  19.0    10.0                         268   
3             2013J             TMA  19.0    10.0                         268   
4             2013J             TMA  19.0    10.0                         268   

  course_id  
0  AAA2013J  
1  AAA2013J  
2  AAA2013J  
3  AAA2013J  
4  AAA2013J

#### Фильтруем по типу Exam, группируем по course_id и id_assessment, заносим в список значения score

In [42]:
course_completeness = studentAssessment_ext.query("assessment_type == 'Exam'") \
    .groupby(['course_id', 'id_assessment'])['score'].apply(list).reset_index(name='score')
course_completeness

course_id  id_assessment                                              score
0  CCC2014B          24290  [32.0, 78.0, 54.0, 64.0, 100.0, 92.0, 84.0, 42...
1  CCC2014J          24299  [44.0, 94.0, 24.0, 70.0, 32.0, 38.0, 68.0, 54....
2  DDD2013B          25340  [40.0, 13.0, 73.0, 40.0, 49.0, 33.0, 51.0, 69....
3  DDD2013J          25354  [69.0, 67.0, 84.0, 33.0, 64.0, 58.0, 44.0, 53....
4  DDD2014B          25361  [58.0, 36.0, 87.0, 73.0, 51.0, 84.0, 91.0, 76....
5  DDD2014J          25368  [60.0, 100.0, 56.0, 67.0, 71.0, 60.0, 82.0, 9....

#### Первым циклом проходимся по course_completeness, делим количество результатов с баллом >= 40 на общее количество результатов, получаем завершаемость. Находим MIN, MAX завершаемость; записываем завершаемость в словарь result. 
#### Вторым циклом выводим курс/экзамен с самой низкой и самой высокой завершаемостью.

In [43]:
completeness_min = 100
completeness_max = 0
result = {}

for row in course_completeness[:course_completeness.shape[0]].itertuples():
    score_passed = list(filter(lambda x: (x >= 40), row.score))
    completeness = round(100 * len(score_passed)/len(row.score))
    if completeness > completeness_max:
        completeness_max = completeness
    if completeness < completeness_min:
        completeness_min = completeness
    result[row.course_id + '/' + str(row.id_assessment)] = completeness

for key in result:
    if result[key] == completeness_min:
        print(f'Completeness MIN: Course/Assessment {key} has completeness {result[key]}.')
    elif result[key] == completeness_max:
        print(f'Completeness MAX: Course/Assessment {key} has completeness {result[key]}.')

Completeness MIN: Course/Assessment DDD2013B/25340 has completeness 84.
Completeness MAX: Course/Assessment DDD2014B/25361 has completeness 93.


#### Ответ: самый сложный курс/экзамен DDD2013B/25340, самый простой курс/экзамен DDD2014B/25361.

### 3. По каждому предмету определи средний срок сдачи экзаменов (под сдачей понимаем последнее успешное прохождение экзамена студентом).

#### Работаем с studentAssessment_ext. 
#### "Под сдачей понимаем последнее успешное прохождение экзамена студентом" -  перефразируем в "под сдачей понимаем успешное прохождение экзамена студентом", так как студент не будет повторно сдавать успешно сданный экзамен. Сделаем фильтрацию по успешной сдаче экзамена, как в 1 задании, сгруппируем по предмету, найдём среднее от даты сдачи экзамена студентом, округлим до целого.

In [44]:
studentAssessment_ext.query("assessment_type == 'Exam' and score >= 40") \
                .groupby('code_module')['date_submitted'].mean().round()

code_module
CCC    239.0
DDD    238.0
Name: date_submitted, dtype: float64

#### Где же предметы AAA, BBB, EEE, FFF, GGG? Их нет.

In [45]:
studentAssessment_ext.query("assessment_type == 'Exam' and code_module in ['AAA', 'BBB', 'EEE', 'FFF', 'GGG']")

Empty DataFrame
Columns: [id_assessment, id_student, date_submitted, is_banked, score, code_module, code_presentation, assessment_type, date, weight, module_presentation_length, course_id]
Index: []

#### Ответ: средний срок сдачи экзаменов по предметам 'CCC' и 'DDD' равен 239 и 238 дней соответственно.

### 4. Выяви самые популярные предметы (ТОП-3) по количеству регистраций на них. А также предметы с самым большим оттоком (ТОП-3).

#### Работаем с studentRegistration

In [46]:
studentRegistration.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                500.0  
1                500.0  
2                 12.0  
3                500.0  
4                500.0

#### Группируем по предмету, считаем количество строк, сортируем, оставляем топ-3.

In [47]:
studentRegistration.groupby('code_module', as_index = False)['date_registration'] \
            .count().sort_values('date_registration', ascending = False).head(3)

code_module  date_registration
1         BBB               7901
5         FFF               7752
3         DDD               6260

#### Топ-3 по оттоку. Фильтруем по дате отмены регистрации != 500. Ранее мы задали значение 500 для не отменённых курсов. Оставляем только отменённые, группируем по предмету, считаем количество строк, сортируем, оставляем топ-3.

In [48]:
studentRegistration.query('date_unregistration != 500').groupby('code_module', as_index = False) \
    ['date_unregistration'].count().sort_values('date_unregistration', ascending = False).head(3)

code_module  date_unregistration
5         FFF                 2370
1         BBB                 2369
3         DDD                 2223

#### Ответ:
#### ТОП-3 предметов по количеству регистраций: BBB, FFF, DDD;
#### ТОП-3 предметов с самым большим оттоком: FFF, BBB, DDD.

### 5. Используя pandas, в период с начала 2013 по конец 2014 выяви семестр с самой низкой завершаемостью курсов и самыми долгими средними сроками сдачи курсов.

#### Это комбинация немного изменённых 2 и 3 заданий
#### Фильтруем по типу Exam, группируем по code_presentation, заносим в список значения score

In [49]:
presentation_completeness = studentAssessment_ext.query("assessment_type == 'Exam'") \
    .groupby('code_presentation')['score'].apply(list).reset_index(name='score')
presentation_completeness

code_presentation                                              score
0             2013B  [40.0, 13.0, 73.0, 40.0, 49.0, 33.0, 51.0, 69....
1             2013J  [69.0, 67.0, 84.0, 33.0, 64.0, 58.0, 44.0, 53....
2             2014B  [32.0, 78.0, 54.0, 64.0, 100.0, 92.0, 84.0, 42...
3             2014J  [44.0, 94.0, 24.0, 70.0, 32.0, 38.0, 68.0, 54....

#### Первым циклом проходимся по presentation_completeness, делим количество результатов с баллом >= 40 на общее количество результатов, получаем завершаемость. Находим MIN завершаемость, записываем завершаемость в словарь result.
#### Вторым циклом выводим семестр с самой низкой завершаемостью.

In [50]:
completeness_min = 100
result = {}

for row in presentation_completeness[:presentation_completeness.shape[0]].itertuples():
    score_passed = list(filter(lambda x: (x >= 40), row.score))
    completeness = round(100 * len(score_passed)/len(row.score))
    if completeness < completeness_min:
        completeness_min = completeness
    result[row.code_presentation] = completeness
    print(f'Presentation: {row.code_presentation}\tCompleteness: {completeness}')

for key in result:
    if result[key] == completeness_min:
        print(f'Completeness MIN: Presentation {key} has completeness {result[key]}.')

Presentation: 2013B	Completeness: 84
Presentation: 2013J	Completeness: 91
Presentation: 2014B	Completeness: 90
Presentation: 2014J	Completeness: 88
Completeness MIN: Presentation 2013B has completeness 84.


In [51]:
studentAssessment_ext.query("assessment_type == 'Exam' and score >= 40") \
                .groupby('code_presentation')['date_submitted'].mean().round().sort_values()

code_presentation
2013B    230.0
2014B    233.0
2013J    240.0
2014J    244.0
Name: date_submitted, dtype: float64

#### Ответ: минимальная завершаемость у семестра 2013B, а максимальный средний срок сдачи курсов у семестра 2014J.

### 6. Часто для качественного анализа аудитории использую подходы, основанные на сегментации. Используя python, построй адаптированные RFM-кластеры студентов, чтобы качественно оценить свою аудиторию. В адаптированной кластеризации можешь выбрать следующие метрики: R - среднее время сдачи одного экзамена, F - завершаемость курсов, M - среднее количество баллов, получаемое за экзамен. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров.

#### Работаем с studentAssessment_ext

In [52]:
studentAssessment_ext.head()

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0           1752       11391              18          0   78.0         AAA   
1           1752       28400              22          0   70.0         AAA   
2           1752       31604              17          0   72.0         AAA   
3           1752       32885              26          0   69.0         AAA   
4           1752       38053              19          0   79.0         AAA   

  code_presentation assessment_type  date  weight  module_presentation_length  \
0             2013J             TMA  19.0    10.0                         268   
1             2013J             TMA  19.0    10.0                         268   
2             2013J             TMA  19.0    10.0                         268   
3             2013J             TMA  19.0    10.0                         268   
4             2013J             TMA  19.0    10.0                         268   

  course_id  
0  AAA2013J  
1  AAA2013J  
2  AAA2013J  
3  AAA2013J  
4  AAA2013J

#### Найдём M - среднее количество баллов, получаемое за экзамен. Фильтруем studentAssessment_ext по сданным экзаменам, группируем по id экзамена, считаем среднее кол-во баллов, округляем до целых, приводим к типу int.

In [53]:
m = studentAssessment_ext.query("assessment_type == 'Exam' and score >= 40") \
    .groupby('id_assessment', as_index=False)['score'].mean().round()
m = m.astype({'score':'int'})
m

id_assessment  score
0          24290     75
1          24299     74
2          25340     61
3          25354     68
4          25361     72
5          25368     68

#### Найдём R - среднее время сдачи одного экзамена. Фильтруем studentAssessment_ext по сданным экзаменам, группируем по id экзамена, считаем среднее время сдачи, округляем до целых, приводим к типу int.

In [54]:
r = studentAssessment_ext.query("assessment_type == 'Exam' and score >= 40") \
    .groupby('id_assessment', as_index=False)['date_submitted'].mean().round()
r = r.astype({'date_submitted':'int'})
r

id_assessment  date_submitted
0          24290             232
1          24299             244
2          25340             230
3          25354             240
4          25361             235
5          25368             243

#### Найдём F - завершаемость курсов
#### Проходимся по course_completeness (создан во 2-м задании), делим количество результатов с баллом >= 40 на общее количество результатов, получаем завершаемость, записываем в словарь result.
#### На основе result создаём датафрейм f, приводим id_assessment к типу int.

In [55]:
result = {}

for row in course_completeness[:course_completeness.shape[0]].itertuples():
    score_passed = list(filter(lambda x: (x >= 40), row.score))
    completeness = round(100 * len(score_passed)/len(row.score))
    result[str(row.id_assessment)] = completeness

f = pd.DataFrame(list(result.items()), columns = ['id_assessment', 'completeness'])
f = f.astype({'id_assessment':'int'})
f

id_assessment  completeness
0          24290            89
1          24299            87
2          25340            84
3          25354            91
4          25361            93
5          25368            89

#### Объединим r, f, m в одну таблицу

In [56]:
rfm = r.merge(f, on = 'id_assessment').merge(m, on = 'id_assessment')
rfm

id_assessment  date_submitted  completeness  score
0          24290             232            89     75
1          24299             244            87     74
2          25340             230            84     61
3          25354             240            91     68
4          25361             235            93     72
5          25368             243            89     68

#### Получили:
* date_submitted - r - среднее время сдачи - чем меньше, тем лучше;
* completeness - f - завершаемость - чем больше, тем лучше;
* score - m - среднее количество баллов - чем больше, тем лучше
#### Проранжируем. Сначала разобьём на 5 интервалов:

In [57]:
quintiles = rfm[['date_submitted', 'completeness', 'score']].quantile([.2, .4, .6, .8]).to_dict()
quintiles

{'date_submitted': {0.2: 232.0, 0.4: 235.0, 0.6: 240.0, 0.8: 243.0},
 'completeness': {0.2: 87.0, 0.4: 89.0, 0.6: 89.0, 0.8: 91.0},
 'score': {0.2: 68.0, 0.4: 68.0, 0.6: 72.0, 0.8: 74.0}}

#### Создадим функции присвоения рангов

In [58]:
def r_score(x):
    if x <= quintiles['date_submitted'][.2]:
        return 5
    elif x <= quintiles['date_submitted'][.4]:
        return 4
    elif x <= quintiles['date_submitted'][.6]:
        return 3
    elif x <= quintiles['date_submitted'][.8]:
        return 2
    else:
        return 1

def fm_score(x, c):
    if x <= quintiles[c][.2]:
        return 1
    elif x <= quintiles[c][.4]:
        return 2
    elif x <= quintiles[c][.6]:
        return 3
    elif x <= quintiles[c][.8]:
        return 4
    else:
        return 5

#### Присвоим ранги с помощью функций r_score() и fm_score(). Сложим ранги-веса, получим RFM_Score, отсортируем по убыванию.

In [59]:
rfm['R'] = rfm['date_submitted'].apply(lambda x: r_score(x))
rfm['F'] = rfm['completeness'].apply(lambda x: fm_score(x, 'completeness'))
rfm['M'] = rfm['score'].apply(lambda x: fm_score(x, 'score'))
rfm['RFM_Score'] = rfm['R'].map(str) + rfm['F'].map(str) + rfm['M'].map(str)
rfm = rfm.sort_values('RFM_Score', ascending = False)
rfm

id_assessment  date_submitted  completeness  score  R  F  M RFM_Score
0          24290             232            89     75  5  2  5       525
2          25340             230            84     61  5  1  1       511
4          25361             235            93     72  4  5  3       453
3          25354             240            91     68  3  4  1       341
5          25368             243            89     68  2  2  1       221
1          24299             244            87     74  1  1  4       114

#### Исходя из RFM_Score можем сделать следующие выводы:
#### - самые успешные экзамены (быстрее сдаются, с высокой завершаемостью и баллами) 24290 и 25361;
#### - самый неуспешный экзамен (медленнее сдаётся, с низкой завершаемостью и баллами) 25368.